src/hotspot/cpu/aarch64/assembler_aarch64.hpp

```diff
@@ -2707,25 +2718,29 @@ template<typename R, typename... Rx>
 #undef INSN
 
 // Advanced SIMD three same
-#define INSN(NAME, op1, op2, op3)                                                       \
-  void NAME(FloatRegister Vd, SIMD_Arrangement T, FloatRegister Vn, FloatRegister Vm) { \
-    starti;                                                                             \
-    assert(T == T2S || T == T4S || T == T2D, "invalid arrangement");                    \
-    f(0, 31), f((int)T & 1, 30), f(op1, 29), f(0b01110, 28, 24), f(op2, 23);            \
-    f(T==T2D ? 1:0, 22); f(1, 21), rf(Vm, 16), f(op3, 15, 10), rf(Vn, 5), rf(Vd, 0);    \
-  }
-
-  INSN(fabd, 1, 1, 0b110101);
-  INSN(fadd, 0, 0, 0b110101);
-  INSN(fdiv, 1, 0, 0b111111);
-  INSN(faddp, 1, 0, 0b110101);
-  INSN(fmul, 1, 0, 0b110111);
-  INSN(fsub, 0, 1, 0b110101);
-  INSN(fmla, 0, 0, 0b110011);
-  INSN(fmls, 0, 1, 0b110011);
-  INSN(fmax, 0, 0, 0b111101);
-  INSN(fmin, 0, 1, 0b111101);
-  INSN(facgt, 1, 1, 0b111011);
+#define INSN(NAME, op1, op2, op3)                                                           \
+  void NAME(FloatRegister Vd, SIMD_Arrangement T, FloatRegister Vn, FloatRegister Vm) {     \
+    starti;                                                                                 \
+    assert(T== T4H || T == T8H || T == T2S || T == T4S || T == T2D, "invalid arrangement"); \
+    int op22 = (T == T2S || T == T4S) ? 0b0 : 0b1;                                          \
+    int op21 = (T == T4H || T == T8H) ? 0b0 : 0b1;                                          \
+    int op15 = (T == T4H || T == T8H) ? 0b00 : 0b11;                                        \
+    f(0, 31), f((int)T & 1, 30), f(op1, 29), f(0b01110, 28, 24), f(op2, 23);                \
+    f(op22, 22); f(op21, 21), rf(Vm, 16), f(op15, 15, 14), f(op3, 13, 10), rf(Vn, 5);       \
+    rf(Vd, 0);                                                                              \
```

Comment:
This is a bit too big for an inline function. Please consider taking the body of this asm and putting it into assembler_aarch64.cpp. See `Assembler::add_sub_immediate` for an example of how this should be done.